Add option to customize imported package names

gonzojive opened the pull request:
Without this, I don't believe it's possible to specify the alias used by a GoFile. This can be important if gopoet would otherwise output an un-aliased import that is not desired.

e.g.

poet was producing

```go
import "x/y/z/blah_go_proto"
```

Such imports are disliked by the lint tool we use.

See related issue #10.

gonzojive committed:
Add (*Imports).RegisterAliasedImport method for forcing a particular import
alias for an import path.

See related issue https://github.com/jhump/gopoet/issues/10.

gonzojive committed:
fix tests for https://github.com/jhump/gopoet/pull/11

jhump commented on go.mod:
```diff
@@ -2,9 +2,4 @@ module github.com/jhump/gopoet
-	golang.org/x/lint v0.0.0-20200302205851-738671d3881b // indirect
-	honnef.co/go/tools v0.0.1-2020.1.4 // indirect
-)
+require github.com/google/go-cmp v0.5.6
```
Please don't commit these removals. This is how the versions are pinned for the static checkers used in CI.

gonzojive replied:
Done.

jhump commented on go.sum:
```diff
@@ -1,42 +1,4 @@
-golang.org/x/tools v0.0.0-20200522201501-cb1345f3a375/go.mod h1:EkVYQZoAsY45+roYkvgYkIh4xh/qjgUK9TdY2XT94GE=
-golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
-golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
+github.com/google/go-cmp v0.5.6 h1:BKbKCqvP6I+rmFHt06ZmyQtvB8xAkWdhFyr0ZUNZcxQ=
```
Do not rewrite go.sum. When you add a new dependency, it should _add_ the checksums for the new stuff. Please don't remove everything that was already here.

gonzojive replied:
Sorry about that. I just ran `go mod tidy` and it did all of the changes to go.mod and go.sum.

jhump commented on imports.go:
```diff
@@ -61,6 +61,39 @@ func (i *Imports) RegisterImport(importPath, packageName string) string {
+// returns the package prefix to use for symbols in the imported package. It is
+// safe to import the same package repeatedly using the same alias.
+//
+// If no
```
Incomplete sentence.

jhump commented on imports.go:
```diff
@@ -61,6 +61,39 @@ func (i *Imports) RegisterImport(importPath, packageName string) string {
+// However, this function will return an error if the given importPath has
+// already been registered and cannot be identified using the provided package
+// name.
+func (i *Imports) RegisterAliasedImport(importPath, packageName string) (string, error) {
```
I think the convention for other functions in this package is to panic if used erroneously.

gonzojive replied:
The reason for the error is that it's hard for the caller to know ahead of time that the given package name or import path is already registered. Most other calls are OK if a package is added with a different alias.

That said, perhaps the API can change in a different way (as mentioned below) instead of adding this method.

gonzojive replied:
Outdated, closing out.

jhump reviewed (commented):
Instead of a new registration method, I think it might instead be better to provide an option (new vararg param) to `ImportSpecs` that tells it to always use an alias if the package name differs from the import path's bae name, and also add that to `WriteGoFile`.

As you see from the checks and possible errors, the import aliases are not intended to be explicitly managed because of the fact that it is easy for the caller to make an error and provide the same path with a different alias or the re-use the same alias for a different import path.

gonzojive committed:
restore go.mod and go.sum

gonzojive committed:
Fix docstring for RegisterAliasedImport.

gonzojive commented:
> Instead of a new registration method, I think it might instead be better to provide an option (new vararg param) to ImportSpecs that tells it to always use an alias if the package name differs from the import path's bae name, and also add that to WriteGoFile.
> 
> As you see from the checks and possible errors, the import aliases are not intended to be explicitly managed because of the fact that it is easy for the caller to make an error and provide the same path with a different alias or the re-use the same alias for a different import path.

A new configuration option instead of the new registration method could work.  What about adding such an option to NewGoFile and NewImportsFor? Or just adding a configuration option or two that can be set on an *Imports object? My concern with adding an option to ImportSpecs is it will not provide enough control - I need to ensure the package name values being generated by Imports are valid.

I did some experimenting here with gopoet but felt it was a bit verbose:
https://github.com/gonzojive/gopoet/commits/fix-10

I also put some import management facilities into a totally separate package:
https://pkg.go.dev/github.com/google/xtoproto@v0.0.16/codegen
In that code, there is a "suggestAliases" method that can be customized (see [here](https://github.com/google/xtoproto/blob/v0.0.16/codegen/codegen_imports.go#L128) and [here](https://github.com/google/xtoproto/blob/75ebe6379ee7599212b46f0c4e41567618fed2d1/codegen/codegen_imports.go#L82)). There is also a [UseAssumedPackageNameFunc](https://pkg.go.dev/github.com/google/xtoproto@v0.0.16/codegen#UseAssumedPackageNameFunc) option that allows specifying the logic used to guess a package name from an import path, and there is also a copy of the logic used by the go tools (which is slightly more nuanced than `path.Base`).

gonzojive committed:
fix go.sum

gonzojive committed:
Add NewGoFile and NewImportsFor option to customize import names.

NewImportsFor now takes a vararg ImportsOption. One such option is
CustomPackageNameSuggester(...) that can be used to control the import specs
printed by an *Imports object.

NewGoFilen ow takes a vararg GoFileOption. The only such option is
GoFileImportsOption(...) which takes options that should be applied to the
embedded Imports object.

gonzojive commented:
The latest commit removes the new method and uses a vararg option to NewGoFile to configure how ImportSpecs and identifier prefixes get generated. Please take a look.

(Relevant commit: https://github.com/jhump/gopoet/pull/11/commits/859c06de4aeb9930b68b5f3ee93f43c215dbb308)

gonzojive committed:
fix typo in CustomPackageNameSuggester docstring

gonzojive commented:
Any comments on the latest revision of the pull request? I'm using it in the version of gopoet at Google, and it's useful for dealing with the linter's quirks.

jhump commented on imports.go:
```diff
@@ -8,16 +8,51 @@ import (
+// stop suggesting package names. The arguments to the callback are the package
+// name to use and whether or not that package name should be considered an
+// alias.
+func CustomPackageNameSuggester(fn func(importPath, packageNameInPackageClause string, callback func(packageName string, isAlias bool) (keepGoing bool))) ImportsOption {
```
This signature is rather complicated and makes a custom package name suggester much more complicated, as it must be able to keep generating suggestions.

Instead, what if the signature simply returned the suggested name (and whether it's an alias). And if that name is not acceptable (due to conflict with another alias), then the `Imports` impl (not the custom suggester) is what is responsible for generating alternatives until a non-conflicting name is found.

If that doesn't work, can you provide some details on what your custom suggester is doing to generate non-conflicting names? Maybe that is a separate aspect that could be customized (instead of having the name suggester _also_ have to handle generation of unique alternatives).

gonzojive replied:
Package names for protobufs must end in 'pb', so if 'blah_go_proto' is the inferred package name, 'blahpb' would be the first alias, 'blah2pb' would be the next, etc. I suspect this would be incompatible with a naive suggester that would append a number in case of a conflict because 'blahpb2' is not desired.

(I would only expect advanced users to use this function since it's optional, so the complexity doesn't concern me.)

jhump replied:
Okay. One thing I think might help: invert the meaning of the callback's result (e.g. `ok` instead of `keepGoing`). I think that leads to more natural code at the point of usage:
```go
if callback(proposal, true) {
    // proposal accepted
    return
}
```

jhump reviewed (commented):
Hello! Sorry for the delays. Thanks for the updates.

This seems like a cool addition. However, I think it has some usability issues. I left another comment detailing my concerns. I'm hoping there's a way to plugin your package name customization in a way that might be easier for others to re-use.

Although I also wonder if this level of flexibility is really necessary. I though the main issue is that you couldn't force `gopoet` to include an import alias (if it thinks that the package's name and the import name are the same, it will omit the alias). And just addressing that would be far simpler than what's in this PR.

jhump commented on imports.go:
```diff
@@ -79,30 +118,73 @@ func (i *Imports) prefixForPackage(importPath, packageName string, registerIfNot
+// imported package." The file being loaded is not available in gopoet (and many
+// go tools), so this function needs to be used.
+func importPathToAssumedName(importPath string) string {
+	// Note: path.Base differs from the package name guesser used by most
```
What does "most tools" mean in this context? The given link is to an internal package, so I don't see how "most tools" could be using it. Or do most tools _fork_ the logic therein? Or is there some exported function that can be used (that delegates to the linked one)? If there's an exported function, I'd be fully in favor of just using it here.

jhump reviewed (commented):
I left a couple of more comments -- one question and one request. Otherwise, I think it LG.
